### PR TITLE
Update energy suppliers

### LINKED ISSUE #333: When metering point master data has been updated the future and current balance supplier is informed about changes.
**Benefit hypothesis**
As a current and future balance supplier
We want to be informed about changes to metering point master data made by the grid company
So that we have the most recent and correct data

**Description**
The purpose of this story is to complete the update metering point master data flow by adding the outgoing message that must be sent to future and current balance supplier when metering point master data has been updated. 

**Acceptance criteria**
- [ ] When metering point master data has been updated, a message is sent to the current and future balance suppliers informing them about the update.
- [ ] The information must be sent to the future and current balance supplier no later than 1 hour after the master data has been updated.  
- [ ] When a update metering point master data request is received, and no changes are made, we must still send a master data changed message to current and future balance suppliers.

**Tasks**
- [x] Eksponer events fra market role. E.g. energy supplier registrered, changed
- [x] Infrastructure
  - Nye topics
  - Subscriptions
- [x] Subscribe on events received from market role
- [x] Data model der kan håndtere el leverandør forhold
- [x] Domæne model der kan håndtere el leverandør forhold
- [x] Command der kan opdatere el leverandør forhold
- [ ] Dannelse af besked (RSM-022)
- [x] logik til håndtere af udsendelse til nuværende og fremtidige elleverandører (husk transaction bøvl)
- [x] Query alt data for MP til besked der skal udsendes
- [ ] <del>Listen to effectuate and annullment integration events</del>

Missing:
- Roles for market participants. From MR?
- "Actor Register"
- Grid operator GLN
- NextReadingDate
- Parent/Child
- Line endings/encoding/BOM?
 
**DoD**

- [ ] All tasks have been completed
- [ ] All acceptance criteria have been completed
- [ ] Unit test performed
- [ ] Integration test performed
- [ ] Tested
- [ ] Accepted by business
- [ ] Solution is deployed to test environment

### PR DESCRIPTION
**Pull request by cmgrd**
## Description
Send RSM 22 messages

Missing:
- Roles for market participants. From MR?
- "Actor Register"
- Grid operator GLN
- NextReadingDate
- Parent/Child
- Line endings/encoding/BOM?

## References

Fixes #333

### CODE REVIEW

File: source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/EnergySupplierDetails.cs [hunk 1]
```diff
@@ -12,23 +12,42 @@
...
 {
     public class EnergySupplierDetails : ValueObject
     {
-        public EnergySupplierDetails(Instant startOfSupply)
+#pragma warning disable CS8618 // Ignore uninitialized properties
+        protected EnergySupplierDetails() { }
+#pragma warning restore
+
+        private EnergySupplierDetails(
+            Guid id,
```
Comment (kft) [inline source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/EnergySupplierDetails.cs:29]: What kind of ID is this ?

File: source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/EnergySupplierDetails.cs [hunk 2]
```diff
@@ -12,23 +12,42 @@
...
+            MeteringPointId marketMeteringPointId,
+            Instant startOfSupply,
+            GlnNumber glnNumber)
```
Comment (kft) [inline source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/EnergySupplierDetails.cs:32]: Do we need GLN-number here ?

File: source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/MarketMeteringPoint.cs [hunk 3]
```diff
@@ -71,7 +73,7 @@ public abstract class MarketMeteringPoint : MeteringPoint
             PowerPlantGsrnNumber = powerPlantGsrnNumber;
         }
 
-        protected EnergySupplierDetails? EnergySupplierDetails { get; private set; }
+        protected Collection<EnergySupplierDetails> EnergySupplierDetails { get; } = new();
```
Comment (kft) [inline source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/MarketMeteringPoint.cs:76]: Is the metering point the right place for this responsibility ?

File: source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/EnergySupplierDetails.cs (hunk 2 above)
Comment (cmgrd) [reply source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/EnergySupplierDetails.cs:32]: It's for recipients of documents. Where should it be?

File: source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/MarketMeteringPoint.cs (hunk 3 above)
Comment (cmgrd) [reply source/Energinet.DataHub.MeteringPoints.Domain/MeteringPoints/MarketMeteringPoints/MarketMeteringPoint.cs:76]: Where should it be?

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/AccountingPointCharacteristics/BusinessReasonCodes.cs [hunk 4]
```diff
@@ -12,11 +12,10 @@
...
 
-namespace Energinet.DataHub.MeteringPoints.Infrastructure.EDI.Common.Address
+namespace Energinet.DataHub.MeteringPoints.Infrastructure.EDI.AccountingPointCharacteristics
 {
-    public record Status(
-        string Value,
-        string DateTime,
-        string Remark,
-        string Reason);
+    public static class BusinessReasonCodes
```
Comment (erduna) [inline source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/AccountingPointCharacteristics/BusinessReasonCodes.cs:17]: Should this be called business reason code? It's actually a process type code

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/AccountingPointCharacteristics/MarketEvaluationPoint.cs [hunk 5]
```diff
@@ -29,9 +29,9 @@ namespace Energinet.DataHub.MeteringPoints.Infrastructure.EDI.AccountingPointCha
         string NetSettlementGroup,
         string NextReadingDate,
         Mrid MeteringGridAreaDomainId,
-        Mrid InMeteringGridAreaDomainId,
-        Mrid OutMeteringGridAreaDomainId,
-        Mrid LinkedMarketEvaluationPoint,
+        Mrid? InMeteringGridAreaDomainId,
+        Mrid? OutMeteringGridAreaDomainId,
+        string LinkedMarketEvaluationPoint,
```
Comment (erduna) [inline source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/AccountingPointCharacteristics/MarketEvaluationPoint.cs:34]: this is actually optional. Comment in master data attribute sheet:

"Skal kun medtages for forbrugs- og produktions-målepunkter, hvor nettoafregning er forskellig fra gruppe 0"

File: source/Energinet.DataHub.MeteringPoints.Tests/Domain/EnergySuppliers/EnergySuppliersTests.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            var sut = DataHub.MeteringPoints.Domain.MeteringPoints.MarketMeteringPoints.EnergySuppliers.Create(Array.Empty<EnergySupplierDetails>());
+
+            var newEnergySupplier = CreateDetails(SampleData.Today, "1");
+            sut.Add(newEnergySupplier);
+
+            sut.GetCurrent(SampleData.Today).Should().Be(newEnergySupplier);
+        }
+
+        [Fact]
+        public void Add_with_existing_on_same_date_should_not_be_ignored()
```
Comment (erduna) [inline source/Energinet.DataHub.MeteringPoints.Tests/Domain/EnergySuppliers/EnergySuppliersTests.cs:44]: This should never be an option, as MR domain will reject such a request.

**Review by erduna — DISMISSED** [review]

File: source/Energinet.DataHub.MeteringPoints.Tests/Domain/EnergySuppliers/EnergySuppliersTests.cs (hunk 6 above)
Comment (cmgrd) [reply source/Energinet.DataHub.MeteringPoints.Tests/Domain/EnergySuppliers/EnergySuppliersTests.cs]: This file is actually not included in the csproj anymore - deleted it!

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/AccountingPointCharacteristics/BusinessReasonCodes.cs (hunk 4 above)
Comment (cmgrd) [reply source/Energinet.DataHub.MeteringPoints.Infrastructure/EDI/AccountingPointCharacteristics/BusinessReasonCodes.cs:17]: In CIM/EDI we're calling these BusinessReasonCodes :-)

**Review by erduna — APPROVED** [review]

**Review by Korgath — APPROVED** [review]

**Pull request merged**